daemon/smartcard.c

```diff
@@ -331,6 +584,137 @@ smartcard_cb_generic(int fd, unsigned events, event_io_t *io, void *data)
+				container_set_token_is_linked_to_device(startdata->container, true);
+				control_send_message(
+					CONTROL_RESPONSE_CONTAINER_CHANGE_PIN_SUCCESSFUL, resp_fd);
+			}
+		} break;
+		case TOKEN_TO_DAEMON__CODE__PROVISION_PIN_FAILED: {
+			container_set_token_is_linked_to_device(startdata->container, false);
+			control_send_message(CONTROL_RESPONSE_CONTAINER_CHANGE_PIN_FAILED, resp_fd);
+		} break;
+		default:
+			ERROR("TokenToDaemon command %d not expected as answer to change_pin",
+			      msg->code);
+		}
+		control_send_message(CONTROL_RESPONSE_CONTAINER_CHANGE_PIN_FAILED, resp_fd);
+		protobuf_free_message((ProtobufCMessage *)msg);
+		event_remove_io(io);
+		event_io_free(io);
+		mem_free(startdata);
+	}
+}
+
+int
+smartcard_change_container_pin(smartcard_t *smartcard, control_t *control, container_t *container,
+			       const char *passwd, const char *newpasswd)
+{
+	ASSERT(smartcard);
+	ASSERT(container);
+	ASSERT(control);
+	ASSERT(passwd);
+	ASSERT(newpasswd);
+
+	int ret = -1;
+	int pw_size = strlen(passwd);
+	int newpw_size = strlen(newpasswd);
+	unsigned char pair_sec[MAX_PAIR_SEC_LEN];
+	int resp_fd = control_get_client_sock(control);
+	bool is_provisioning;
+
+	smartcard_startdata_t *startdata = mem_alloc(sizeof(smartcard_startdata_t));
+	if (!startdata) {
+		ERROR("Could not allocate memory for startdata");
+		return -1;
+	}
+	startdata->smartcard = smartcard;
+	startdata->container = container;
+	startdata->control = control;
+
+	DEBUG("SCD: Passwd form UI: %s, size: %d", passwd, pw_size);
+	DEBUG("SCD: New Passwd form UI: %s, size: %d", newpasswd, newpw_size);
+
+	ret = smartcard_get_pairing_secret(smartcard, pair_sec, sizeof(pair_sec));
+	if (ret < 0) {
+		ERROR("Could not retrieve pairing secret, ret code : %d", ret);
+		control_send_message(CONTROL_RESPONSE_CONTAINER_CHANGE_PIN_FAILED, resp_fd);
+		return -1;
+	}
+
+	is_provisioning = !smartcard_container_token_is_provisioned(container);
+
+	event_io_t *event = event_io_new(smartcard->sock, EVENT_IO_READ,
+					 smartcard_cb_change_container_pin, startdata);
+	event_add_io(event);
+	DEBUG("SCD: Registered generic container callback for scd");
+
+	DaemonToToken out = DAEMON_TO_TOKEN__INIT;
+	out.code = is_provisioning ? DAEMON_TO_TOKEN__CODE__PROVISION_PIN :
+				     DAEMON_TO_TOKEN__CODE__CHANGE_PIN;
+
+	out.token_uuid = mem_strdup(uuid_string(container_get_uuid(startdata->container)));
+
+	out.has_token_type = true;
+	out.token_type =
+		smartcard_tokentype_to_proto(container_get_token_type(startdata->container));
+
+	out.token_pin = mem_strdup(passwd);
+	out.token_newpin = mem_strdup(newpasswd);
+
+	out.has_pairing_secret = true;
+	out.pairing_secret.len = sizeof(pair_sec);
+	out.pairing_secret.data = mem_memcpy(pair_sec, sizeof(pair_sec));
+
+	ret = protobuf_send_message(smartcard->sock, (ProtobufCMessage *)&out);
+	mem_free(out.token_pin);
+	mem_free(out.token_newpin);
+	mem_free(out.pairing_secret.data);
+	mem_free(out.token_uuid);
+
+	return (ret > 0) ? 0 : -1;
```

Comment:
>= 0 (see protobuf_send_message, 0 is not an error)

Reply:
but in this case it would actually be, because an all-default message is not possible here